src/vs/vscode.d.ts

```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
 	}
 
 	/**
+	 * A node provider for the tree explorer contributed by extension.
+	 *
+	 * Providers are registered through (#workspace.registerTreeExplorerNodeProvider) with a
+	 * `providerId` that corresponds to the `treeExplorerNodeProviderId` in the extension's
+	 * `contributes.explorer` section.
+	 *
+	 * The contributed tree explorer will ask the corresponding provider to provide the root
+	 * node and resolve children for each node. In addition, the provider could **optionally**
+	 * provide the following information for each node:
+	 * - label: A human-readable label used for rendering the node.
+	 * - hasChildren: Whether the node has children and is expandable.
+	 * - clickCommand: A command to execute when the node is clicked.
+	 */
+	export interface TreeExplorerNodeProvider<T> {
```

Comment:
You don't need to up-cast anything when your define it as `<T extends TreeItem>`. It makes an easy to understand API and *most importantly* it aligns with other parts of the API. Keeping things consistent (look at quick pick) is key to make an easily understandable API. Your sample proofs that the *label provider* indirection is artificial as the actual data is stored in the [node](https://github.com/octref/explorer-api-example/blob/master/dep/src/extension.ts#L34) and not computed by a provider.